src/Storages/IStorage.cpp

```diff
@@ -133,6 +133,13 @@ void IStorage::read(
     size_t num_streams)
 {
     auto pipe = read(column_names, storage_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
+
+    /// parallelize processing if not yet
```

Comment:
1. Should we do it here, or is it better to do it inside `InterpreterSelectQuery`?
2. Are there any potential troubles with mutations and StorageFromMergeTreeDataPart?

Reply:
> 1. Should we do it here, or is it better to do it inside `InterpreterSelectQuery`?

I think it's ok to do it here with the following considerations ...

`num_streams` is provided by `InterpreterSelectQuery` as a recommendation, i.e. how many threads are available for data processing. The reading step has the following choices: 

- (a) it knows the amount of data it can read, and it's not much data, so it creates only the necessary number of data streams based on parameters passed to `IStorage::read()` i.e. `max_block_size`/`storage_limits`. In this case, we don't want to adjust the number of streams and `parallelizeOutputAfterReading()` can return `false`

- (b) it's either an unknown amount of data or known amount of data (but enough to utilize all available threads) -> in both cases output is parallelized by `num_streams` 


> 2. Are there any potential troubles with mutations and StorageFromMergeTreeDataPart?

The generic thing about this change affects only storage which will use default plan step to read from storage – `ReadFromStorageStep`. Sophisticated engines use specialized steps to read from its storage, like `ReadFromMergeTree` in `MergeTree` case.

`StorageFromMergeTreeDataPart` is not affected since it uses the `ReadFromMergeTree` step, which overrides `read()` method where `resize()` is added.